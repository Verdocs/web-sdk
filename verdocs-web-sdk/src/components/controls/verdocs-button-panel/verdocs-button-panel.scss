@import '../../../theme.scss';

verdocs-button-panel {
  font-family: $verdocs-primary-font;
  display: inline-block;

  .icon {
    display: inline-block;
    cursor: pointer;

    svg {
      fill: $light-border-color;
    }

    &:hover {
      svg {
        fill: $verdocs-grey-4;
      }
    }
  }
}

.verdocs-button-panel-content {
  font-family: $verdocs-primary-font;
  box-shadow: 0 0 10px 1px #999999;
  display: none;
  background: #ffffff;
  color: $label-color;
  width: 280px;
  font-weight: bold;
  padding: 15px;
  font-size: 14px;
  position: relative;
  border-radius: 4px;
  z-index: 10000;

  .arrow,
  .arrow::before {
    position: absolute;
    width: 8px;
    height: 8px;
    background: inherit;
  }

  .arrow {
    visibility: hidden;
    top: -4px;
  }

  .arrow::before {
    visibility: visible;
    content: '';
    transform: rotate(45deg);
  }

  &[data-show] {
    display: block;
  }

  &[data-popper-placement^='top'] > .arrow {
    bottom: -4px;
  }

  &[data-popper-placement^='bottom'] > .arrow {
    top: -4px;
  }

  &[data-popper-placement^='left'] > .arrow {
    right: -4px;
  }

  &[data-popper-placement^='right'] > .arrow {
    left: -4px;
  }

  h6 {
    color: $verdocs-grey-0;
    font-size: 16px;
    font-weight: bold;
    margin: 0 0 8px 0;
  }

  form {
    margin: 0;
    padding: 0;
    display: flex;
    flex-direction: column;
  }

  label {
    display: block;
    font-size: 14px;
    font-weight: bold;
    margin: 0 0 3px 0;
    color: $verdocs-grey-1;
  }

  input {
    flex: 1;
    height: 34px;
    display: flex;
    font-size: 14px;
    margin: 0 0 10px 0;
  }
}
