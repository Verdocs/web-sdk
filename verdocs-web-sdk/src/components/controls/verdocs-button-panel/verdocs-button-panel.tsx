import {createPopper, Instance} from '@popperjs/core';
import {Component, h, Host, Method, Prop} from '@stencil/core';

/**
 * Displays a clickable icon suitable for display in a toolbar. When clicked, a customizable drop-down panel will be
 * displayed.
 *
 * *
 * ```html
 * <verdocs-button-panel icon="<svg.../>">
 *   <form>...</form>
 * </verdocs-button-panel>
 * ```
 */
@Component({
  tag: 'verdocs-button-panel',
  styleUrl: 'verdocs-button-panel.scss',
  shadow: false,
})
export class VerdocsButtonPanel {
  iconEl: HTMLDivElement;
  panelEl: HTMLDivElement;
  hiderEl?: HTMLDivElement;
  popperInstance: Instance;

  /**
   * SVG icon to display
   */
  @Prop() icon: string = '';

  componentDidRender() {
    document.body.appendChild(this.panelEl);
    if (this.popperInstance) {
      this.popperInstance.destroy();
    }

    this.popperInstance = createPopper(this.iconEl, this.panelEl, {
      // placement: 'top-end',
      // modifiers: [{name: 'offset', options: {offset: [0, 10]}}],
    });
  }

  disconnectedCallback() {
    if (this.popperInstance) {
      this.popperInstance.destroy();
      this.popperInstance = null;
    }

    if (this.panelEl) {
      this.panelEl.remove();
    }
  }

  // Not marked as @State because it's not the thing that controls our rendering
  showing = false;

  @Method()
  async showPanel() {
    if (!this.showing) {
      await this.toggle();
    }
  }

  @Method()
  async hidePanel() {
    if (this.showing) {
      await this.toggle();
    }
  }

  @Method()
  async toggle() {
    if (this.showing) {
      this.panelEl?.removeAttribute('data-show');
      this.hiderEl?.remove();
      this.showing = false;
    } else {
      this.panelEl?.setAttribute('data-show', '');
      this.popperInstance?.update().catch(() => {});
      this.showing = true;

      this.hiderEl = document.createElement('div');
      this.hiderEl.style.zIndex = '100';
      this.hiderEl.style.position = 'absolute';
      this.hiderEl.style.top = '0px';
      this.hiderEl.style.left = '0px';
      this.hiderEl.style.right = '0px';
      this.hiderEl.style.bottom = '0px';
      this.hiderEl.onclick = (e: any) => {
        e.stopPropagation();
        this.toggle();
      };
      document.body.appendChild(this.hiderEl);
    }
  }

  render() {
    return (
      <Host>
        <div
          class="icon"
          innerHTML={this.icon}
          onClick={(e: any) => {
            e.stopPropagation();
            return this.toggle();
          }}
          ref={el => (this.iconEl = el)}
        />
        <div role="tooltip" class="verdocs-button-panel-content" data-popper-placement="bottom" ref={el => (this.panelEl = el as HTMLDivElement)}>
          <div data-popper-arrow="true" class="arrow" />
          <slot />
        </div>
      </Host>
    );
  }
}
