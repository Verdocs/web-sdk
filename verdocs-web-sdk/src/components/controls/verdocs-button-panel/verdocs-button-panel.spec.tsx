import { newSpecPage } from '@stencil/core/testing';
import { VerdocsButtonPanel } from './verdocs-button-panel';

describe('verdocs-button-panel', () => {
  it('renders with default props', async () => {
    const page = await newSpecPage({
      components: [VerdocsButtonPanel],
      html: `<verdocs-button-panel></verdocs-button-panel>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with icon prop', async () => {
    const page = await newSpecPage({
      components: [VerdocsButtonPanel],
      html: `<verdocs-button-panel icon="<svg></svg>"></verdocs-button-panel>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with slotted content', async () => {
    const page = await newSpecPage({
      components: [VerdocsButtonPanel],
      html: `<verdocs-button-panel icon="<svg></svg>"><div>Panel Content</div></verdocs-button-panel>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  // Note: Panel show/hide is managed via DOM and Popper, which may not be fully testable in JSDOM.
  // This test checks if the icon element exists and is clickable.
  it('has icon element for interaction', async () => {
    const page = await newSpecPage({
      components: [VerdocsButtonPanel],
      html: `<verdocs-button-panel icon="<svg></svg>"></verdocs-button-panel>`,
    });
    const iconDiv = page.root.querySelector('div');
    expect(iconDiv).not.toBeNull();
  });
});
