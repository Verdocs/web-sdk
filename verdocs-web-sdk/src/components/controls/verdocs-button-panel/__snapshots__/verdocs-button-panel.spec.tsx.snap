// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verdocs-button-panel renders with default props 1`] = `
<verdocs-button-panel>
  <!---->
  <div class="icon"></div>
</verdocs-button-panel>
`;

exports[`verdocs-button-panel renders with icon prop 1`] = `
<verdocs-button-panel icon="<svg></svg>">
  <!---->
  <div class="icon">
    <svg></svg>
  </div>
</verdocs-button-panel>
`;

exports[`verdocs-button-panel renders with slotted content 1`] = `
<verdocs-button-panel icon="<svg></svg>">
  <!---->
  <div class="icon">
    <svg></svg>
  </div>
</verdocs-button-panel>
`;
